#81 progress bar logic

nmadruga opened the pull request:
Fixes #81 

~~Needs to run using [Back-end#120](https://github.com/openmsupply/application-manager-server/pull/120) and [Back-end#121](https://github.com/openmsupply/application-manager-server/pull/121) - using the Back-end#121 will have both changes.~~ Back-end issue already merged in.

Use [Back-end-PR#127](https://github.com/openmsupply/application-manager-server/pull/127) to see the Company registration with more pages per section. This will make the test of the Progress bar logic for sections/pages easier.

A brief of the logic for the Progress bar is described in the issue related to this PR, if further discussion regarding the logic it would be good to add directly in there (in the comments) so we can keep it up-to-date.

I wasn't sure how to structure this object that is passed to the ProgressBar component, so any improvements ideas will be very welcome. For now it doesn't show the progress after you move from the Summary back to the sections... But to be able to do something like that, the structure for the ProgressBar needs to be a shared state instead of a local one in the `ApplicationPageWrapper`.

### Changes
- Create 3 types `ProgressInApplication`, `ProgressInPage`, `ProgressInSection` used in hierarchical order to keep the current progress in Application (stored in local state in the `ApplicationPageWrapper`.
- The `ProgressBar` and `NavigationBox` will receive the `validateCurrentPage` function in the props to run the validation on the current page, store in the `progressInApplication` state on the ApplicationPage and return if can proceed to next page.
- Some logic was removed from the `ProgressBar` that now will just present the icons, disable/enable sections depending on what is passed in the `progressInApplication` structure.

### Remaining todo
- Use the validation of fields in the `progressInApplication` structure (is only using `isRequired` and `isVisible` for now)
- On linear Application need to disable next sections (that haven't been visited) on the progress bar.
- Display validation on Progress bar after coming back from the Summary page.

### Screenshots
The progress bar on the Application: Company Registration
* To create one company registration application go to `http://localhost:3000/applications/new?type=CompRego1` and start a new one. It will have 3 sections with one page in each, so when moving to the next pages the whole section is validated.
<img width="716" alt="Screen Shot 2020-11-19 at 12 36 14 PM" src="https://user-images.githubusercontent.com/16461988/99607146-127fc800-2a70-11eb-8fb6-29b956104a87.png">
 
* **Designed in Miro**:
<img width="806" alt="Screen Shot 2020-11-19 at 2 01 40 PM" src="https://user-images.githubusercontent.com/16461988/99607055-de0c0c00-2a6f-11eb-9b12-9578d3545514.png">

CarlosNZ commented:
Now that I see it in action, my initial comment would be that the Progress bar _should_ show up on the Summary page (at least before it's submitted, not after). Otherwise, you never actually see the "All done" icon for the last page/all the pages at once, which is kind of unsatisfying from a user POV. ;)

CarlosNZ reviewed (commented):
Just going to make functionality observations first, haven't really looked at the code yet:

- When going back using the "Previous" button, the progress bar is fine, but when going back using Browser back (or direct to a URL, or clicking in the Application list), the progress bar no longer has the correct/warning icons.
- It seems weird not having the progress bar on the summary page, IMHO.
- Wouldn't it be easier to make the "neutral" icon a generic one rather than numbers? I think we talked about not bothering to number the Sections as you can just put a number in the section name if you want it.

andreievg reviewed (changes requested):
Hi @nmdruga,

I've reviewed `ProgressBars.tsx`, and then moved on to `ApplicationWrapper`, had a quick look at the logic there. I think it's sound, but it's a bit hard to follow, what all of the set, update and validate methods do and how they behave together.

My original thought for the process of validating pages and displaying them was as follow:

(in ApplicationWrapper)

`useGetResponseAndElementState` <- happens when new value (and now isValid) is stored in cache, this one calculates isVisibility etc, return one object with all info for a response/element 

`useEffect on when elementsState has change` <- here we can calculate, base on return from `useGetResponseAndElementState` validity and return an create an object, something like this:

```
[
    {
        sectionName: 'Section 1',
        status: 'valid'|'inValid'|'undefined',
        canNavigate: true|false,
        isActive: true|false,
        pages: [
            {
                pageName: 'Page 1',
                status: 'valid'|'inValid'|'undefined',
                canNavigate: true|false,
                isActive: true|false,
            }
        ]

    }
]
```

The process above kind of automatically deals with all of the use case (i.e. reloading the application etc..).

What do you think ?

nmadruga commented:
Thanks for the feedbacks @carl and @andrei much appreciated!

I'll just comment briefly on the overall reviews - and will be working on some changes for the suggestions you gave next.

nmadruga commented:
> When going back using the "Previous" button, the progress bar is fine, but when going back using Browser back (or direct to a URL, or clicking in the Application list), the progress bar no longer has the "complete"/"warning" icons.

Good point! That's a problem with the current way I have implemented this, with the local state on the `ApplicationWrapper` then anytime it refreshes is going to the initial state. Not good -> I'll be changing this structure to be returned by the hook `useGetResponseAndElementState` as suggested.

> It seems weird not having the progress bar on the summary page, IMHO.

Yes, I was also a bit annoyed by that, hmmm maybe we could have it sort of hidden? I'll try to illustrate this idea on Monday.

> Wouldn't it be easier to make the "neutral" icon a generic one rather than numbers? I think we talked about not bothering to number the Sections as you can just put a number in the section name if you want it.

I think we should go with what has been designed otherwise we aren't considering the inputs from the designer.

nmadruga commented:
> My original thought for the process of validating pages and displaying them was as follow:

> (in ApplicationWrapper)

> useGetResponseAndElementState <- happens when new value (and now isValid) is stored in cache, this one calculates isVisibility etc, return one object with all info for a response/element

>useEffect on when elementsState has change <- here we can calculate, base on return from useGetResponseAndElementState validity and return an create an object, something like this:
> ...

I agree its confusing and I totally agree it needs some changes (move to a hook **or** use a context to store the shared state **or** else), I'm not 100% sure that using a `useEffect` to build this object is the best solution. But I'll be giving a try on this solution, let's see if it can make this simpler. : )

I'm already using the result from `useGetResponseAndElementState` to generate the structure that is sent to the ProgressBar component, but the problem with it storing this object here is that the object is reset when the page refreshes, which might be fixed with this solution since it will be generating the object everytime. Let's see!

nmadruga commented on src/components/Application/ProgressBar.tsx:
```diff
@@ -9,34 +9,55 @@ interface SectionPage {
+                validateCurrentPage,
+              })
+            }
+            active={currentPage.visited} // TODO: Change to only show active when visited in linear applications
```
Yes, you are right. I'll change this. Ty!

nmadruga commented on src/components/Application/ProgressBar.tsx:
```diff
@@ -9,34 +9,55 @@ interface SectionPage {
+            }
+            active={currentPage.visited} // TODO: Change to only show active when visited in linear applications
+          >
+            {currentPage.pageStatus == true ? (
```
Yes, that's what I was aiming with this `pageStatus` and `sectionStatus`, meaning:
`true` => All (required) responses are valid

nmadruga replied:
I'm changing this property to `status` as you suggested on the previous comment. And considering creating a type to store the 3 cases for each page validation: `valid`, `invalid` or `incomplete` => Which would be the status when only a part of the **required responses** have been answered regardless of valid or invalid. What do you think?

nmadruga commented on src/components/Application/ProgressBar.tsx:
```diff
@@ -45,20 +66,35 @@ const ProgressBar: React.FC<ProgressBarProps> = ({
             <Grid>
               <Grid.Column width={4} textAlign="right" verticalAlign="middle">
-                {getStepNumber(stepNumber)}
+                {section.sectionStatus !== undefined ? (
```
I'll will be working on this one now.

nmadruga commented:
Ok, I think this is ready for a re-review... @CarlosNZ @andreievg 

I could set the progress of sections not visited (after the current) to not display the statue on the linear application:
<img width="1055" alt="Screen Shot 2020-11-23 at 3 21 12 PM" src="https://user-images.githubusercontent.com/16461988/99924548-aa96ed80-2d9f-11eb-82c2-4d3038e782a5.png">

The non-linear, as discussed on the planning meeting on **23/11/2020**,  not having a shared local state to determine what sections have been visited and which didn't makes the validation show over the ones which have been visited from the start, for example if there is **no required field** in a page, and it hasn't been filled -> the page is valid.

CarlosNZ commented:
Just noticed:
![Screen Shot 2020-11-23 at 4 21 37 PM](https://user-images.githubusercontent.com/5456533/99926850-fc437600-2da7-11eb-96d3-5d83d5c0797e.png)
-- Sections appearing out of order?

Also -- isn't there meant to be valid/invalid icons for the sections once we've filled them in?
![Screen Shot 2020-11-23 at 4 24 16 PM](https://user-images.githubusercontent.com/5456533/99927026-acb17a00-2da8-11eb-9c9a-c8ee5f7f57aa.png)

CarlosNZ commented:
![Screen Shot 2020-11-23 at 4 29 34 PM](https://user-images.githubusercontent.com/5456533/99927139-1467c500-2da9-11eb-92cc-2ce821912010.png)
Also -- it's odd that Page 1 has an active link when I'm on page 1, and page 2 doesn't (and I've been there before)
(Sorry, the cursor doesn't show in the screenshot as I expected -- but "Page 1" is clickable and "Page 2" isn't.)

nmadruga commented:
Hi @CarlosNZ thanks for having another look into this PR : )

> -- Sections appearing out of order?

Yes, that's a problem there. I had already worked on a fix - added to a separated PR (but forgot to create yesterday) here: #134.

>  Also -- isn't there meant to be valid/invalid icons for the sections once we've filled them in?

Yes. It should do this. There is already logic to do that, for some reason not working right now. I'll have a look after morning break : ) I also want to make sure I'm following the rules we set for Progress bar in the description of issue #81.

> Also -- it's odd that Page 1 has an active link when I'm on page 1, and page 2 doesn't (and I've been there before)
(Sorry, the cursor doesn't show in the screenshot as I expected -- but "Page 1" is clickable and "Page 2" isn't.)

Hmmmm. Not sure how to fix that exactly. I made the `User registration` as a `linear` application, which means you can only move forward (to **Page 2**) once all required fields in **Page 1** are filled and valid. That's why the next page in the progress bar is not clickable, the same if there were more sections further they would be all unclickable.

Not sure if the current page should or shouldn't be clickable though. What do you think? Not clickable?

CarlosNZ commented:
> Not sure if the current page should or shouldn't be clickable though. What do you think? Not clickable?

Yeah, I definitely think the current page should be not clickable. But can make it a new issue if it's getting to complicated in this one.

One other suggestion -- do you think we should add additional pages to the Company Registration template? I feel like I'm not able to fully check the functionality of a non-linear application with more than one page in a section.

nmadruga commented:
> Yeah, I definitely think the current page should be not clickable. But can make it a new issue if it's getting to complicated in this one.

It's ok. I'll add this as part of this one. Shouldn't take me too long.

> One other suggestion -- do you think we should add additional pages to the Company Registration template? I feel like I'm not able to fully check the functionality of a non-linear application with more than one page in a section.

Ok. I'll add more pages to it on another Back-end PR.

nmadruga commented:
This PR is ready for another review : )

I have reviewed the progress states expected in the issue, and linked other issues to be dealing with more specific changes - not part of this PR.

andreievg commented on src/containers/Application/ApplicationPageWrapper.tsx:
```diff
@@ -44,10 +57,42 @@ const ApplicationPageWrapper: React.FC = () => {
+      currentPage: Number(page),
+    })
+    setProgressInApplication(progressStructure)
+    setProcessingValidation(false)
```
Hmm I am not sure why `processingValidation` is required

nmadruga replied:
It was required to call the `ProgressBar` component when the structure have finish the creation. It doesn't seem to be a problem to remove it now, so I've removed [here](https://github.com/openmsupply/application-manager-web-app/pull/127/commits/2336794c474455557585527bfbfde6221859600e)

andreievg replied:
thanks

andreievg commented on src/containers/Application/ApplicationPageWrapper.tsx:
```diff
@@ -44,10 +57,42 @@ const ApplicationPageWrapper: React.FC = () => {
+  }, [isApplicationLoaded, sectionCode, page])
+
+  useEffect(() => {
+    const progressStructure = buildProgressInApplication({
```
`comment`, with multiple effects, it would be good to name them somehow, a quick comment with their intended function would be great.

nmadruga replied:
No problem. I'll do that now

andreievg commented on src/containers/Application/ApplicationPageWrapper.tsx:
```diff
@@ -44,10 +57,42 @@ const ApplicationPageWrapper: React.FC = () => {
+      responses: responsesFullByCode as ResponsesFullByCode,
+      sectionIndex: currentSection?.index as number,
+      page: Number(page),
+      checkEmpty: true,
```
what is the use case of checkEmpty parameter, it's not apparent from reading the code

nmadruga replied:
The `checkEmpty` is used for the **strict** validation. The default then is **loose** validation, which means that the validatePage is not checking for empty responses. I'll move this to a helper file and add some comments to this.

nmadruga replied:
Oh, sorry, it's already in a helper file. I'll just add a comment explaining what I just did here. Then we can improve this function as needed.

andreievg commented on src/utils/helpers/validatePage.tsx:
```diff
@@ -0,0 +1,113 @@
+    .reduce((pageElements: ElementState[], element) => {
+      if (element.elementTypePluginCode === 'pageBreak') count++
+      if (count !== page) return pageElements
+      if (element.category === TemplateElementCategory.Question) return [...pageElements, element]
```
not a required change, this is stylistic preference, but to me logically it makes more sense to check in this fashion:

if pageBreak increment
if not question type category return pageElement
if current page return pageElements + element
return pageElements

nmadruga replied:
Thanks. I have changed it slightly to what you have suggested: [here](https://github.com/openmsupply/application-manager-web-app/pull/127/commits/bd0a16c3cc244a73542f8880aa206894619243cc)

andreievg commented on src/utils/helpers/validatePage.tsx:
```diff
@@ -0,0 +1,113 @@
+
+import { TemplateElementCategory } from '../../utils/generated/graphql'
+
+export const PROGRESS_STATUS: { [key: string]: ProgressStatus } = {
```
i think this is use case for enum ?

andreievg commented on src/utils/helpers/validatePage.tsx:
```diff
@@ -0,0 +1,113 @@
+      return pageElements
+    }, [])
+
+  const responsesStatuses = elementsInCurrentPage.reduce(
```
Just a preliminary question before looking at code below in depth, don't we just want to do this:
(btw nested conditional operators are hard to read)

(considering that elementsInCurrentPage have responses already linked)

filter elementsInCurrentPage by isVisible visibleElementsInPage
filter visibleElementsInPage by isRequired let's call it requiredAndVisibleElementsInPage

if(visibleElementsInPage find isValid === false) then NOT_VALID
if(requiredAndVisibleElementsInPage find by value === undefined ) then INCOMPLETE

otherwise VALID

nmadruga replied:
> filter elementsInCurrentPage by isVisible visibleElementsInPage

nmadruga replied:
> otherwise VALID

There is just one problem with that: we can have notRequired elements that have been filled and are INVALID

andreievg replied:
I think that's good, if something in the form is entered incorrectly (even though it's not required), I think it should be indicated as incorrect

nmadruga replied:
I added more comments should be easier to read now. Thanks for all the help with this function much appreciated! Commit [here](https://github.com/openmsupply/application-manager-web-app/pull/127/commits/4cf4f1b3a292fc1359dd0a99a817faf237c0babc)

andreievg reviewed (changes requested):
@nmadruga Partial Review, still have `useGetResponsesAndElementState` to review and UI

andreievg commented on src/containers/Application/ApplicationPageWrapper.tsx:
```diff
@@ -111,4 +164,83 @@ function processRedirect(appState: any): void {
+  if (!elementsState || !responses) return []
+  let sectionsStructure: ProgressInApplication = templateSections.map((section) => {
+    // Create an array with all pages in each section
+    const pages = Array.from(Array(section.totalPages).keys(), (n) => n + 1)
```
looks pretty cool.

andreievg commented on src/containers/Application/ApplicationPageWrapper.tsx:
```diff
@@ -111,4 +164,83 @@ function processRedirect(appState: any): void {
+    const pages = Array.from(Array(section.totalPages).keys(), (n) => n + 1)
+    const checkIsCurrentPage = (page: number) =>
+      section.index === currentSection && page === currentPage
+    const checkIsPreviousPage = (index: number, page: number) =>
```
don't really need the `check` bit in the method naming here, but doesn't hurt to keep

nmadruga replied:
I'll remove. : )

andreievg commented on src/containers/Application/ApplicationPageWrapper.tsx:
```diff
@@ -111,4 +164,83 @@ function processRedirect(appState: any): void {
+      title: section.title,
+      canNavigate: isLinear ? section.index <= currentSection : true,
+      isActive: section.index === currentSection,
+      pages: pages.map((number) => ({
```
pageNumber ?

andreievg commented on src/containers/Application/ApplicationPageWrapper.tsx:
```diff
@@ -111,4 +164,83 @@ function processRedirect(appState: any): void {
+    const pages = Array.from(Array(section.totalPages).keys(), (n) => n + 1)
+    const checkIsCurrentPage = (page: number) =>
+      section.index === currentSection && page === currentPage
+    const checkIsPreviousPage = (index: number, page: number) =>
```
sectionIindex ?

andreievg commented on src/containers/Application/ApplicationPageWrapper.tsx:
```diff
@@ -111,4 +164,83 @@ function processRedirect(appState: any): void {
+    const checkIsCurrentPage = (page: number) =>
+      section.index === currentSection && page === currentPage
+    const checkIsPreviousPage = (index: number, page: number) =>
+      index < currentSection ? true : index === currentSection && page <= currentPage
```
i see you quite like conditional operators. typically we would just add brackets around `index === currentSection && page <= currentPage` with an or in between.

I think what we want for canNavigate logic, for linear (does not need to be aware of current or previous page):

the first not valid or incomplete page from the start = set flag for cant navigate, then every page and section after that in linear mode, can't be navigated to

nmadruga replied:
Yes, the conditional helps me to think, but I agree is not the easiest to read. I'll change to what you suggest.
I will also consider what you comment on the status for `canNavigate` flag.

nmadruga replied:
> the first not valid or incomplete page from the start = set flag for cant navigate

I think this would be more complicated since we can have any page with status INVALID, that's why I'm checking first for `isPreviousPage`. Although if we want to display all validation in the progress bar after the user has been to the `Summary section` I don't think this code would work. But I'm not sure how we can change it to consider this overall validation.

andreievg replied:
Doesn't `canNavigate` only apply to linear mode? So in linear mode you can only navigate to the first page that is invalid/incomplete ?

andreievg commented on src/containers/Application/ApplicationPageWrapper.tsx:
```diff
@@ -111,4 +164,83 @@ function processRedirect(appState: any): void {
+    const progressInSection: ProgressInSection = {
+      code: section.code,
+      title: section.title,
+      canNavigate: isLinear ? section.index <= currentSection : true,
```
canNavigate to next section if currentSection last page is Valid in linear mode

andreievg commented on src/containers/Application/ApplicationPageWrapper.tsx:
```diff
@@ -111,4 +164,83 @@ function processRedirect(appState: any): void {
+        pageName: `Page ${number}`,
+        canNavigate: isLinear ? checkIsPreviousPage(section.index, number) : true,
+        isActive: checkIsCurrentPage(number),
+        status: checkIsCurrentPage(number)
```
can you comment, what this logic is here, shouldn't it just be :`validatePage(...)` ?

I would do pages outside of the progressInSection object like this:

```

const pages = ...

progressInSection: {
..
status: getCombinedStatus(pages),
pages
}

```

nmadruga replied:
Moved to outside of the scope as suggested: [here](https://github.com/openmsupply/application-manager-web-app/pull/127/commits/b6e437b958e873fd0182e53fa8fa4174cc4a91c9)
 
The reason why the `validatePage` is done 2 ways is differing if it would run in a `strict` or `loose` mode.
1. `linear` and `previous page` it would run a STRICT validation
2. `non-linear` and `previous page` is should run a LOOSE validation

Now in case the user has been to the `Summary` page I think we would need to run a STRICT validation over all the application, which is not implemented yet. I think it would be better if we leave this to be dealt in the issue #48

andreievg commented on src/utils/helpers/validatePage.tsx:
```diff
@@ -0,0 +1,113 @@
+
+export const getCombinedStatus = (
+  array: { status: ProgressStatus }[] | undefined,
+  strictly: boolean = false
```
what is strictly for this method ?

nmadruga replied:
I was reconsidering this one. I think this is not needed anymore. I was thinking we would show the section state (combined of pages statuses) different when we ran the validation. After our meetings and the document that @CarlosNZ created that the combined status would always return:
> Any INVALID page in Section → Section state = ❌
> Any NEUTRAL page in Section → Section state = Neutral
> ELSE:  Section state = ✅

andreievg commented on src/utils/helpers/validatePage.tsx:
```diff
@@ -0,0 +1,113 @@
+}
+
+export const getCombinedStatus = (
+  array: { status: ProgressStatus }[] | undefined,
```
I think it's much simpler to just do `some` here:

pages.some(({status} )=> status === NOT_VALID) return NOT_VALID
pages.some(({status} )=> status === INCOMPLETE) return INCOMPLETE
return VALID

nmadruga replied:
Yes! I agree (only saw this after I commented on the previous one).
Change [here](https://github.com/openmsupply/application-manager-web-app/pull/127/commits/c55bfbf478f1f7b772b0ce4a506fe4cbb246e5b8)

andreievg reviewed (changes requested):
I think only UI remains now

andreievg commented on src/containers/Application/ApplicationPageWrapper.tsx:
```diff
@@ -44,10 +57,42 @@ const ApplicationPageWrapper: React.FC = () => {
+  }, [isApplicationLoaded, sectionCode, page])
+
+  useEffect(() => {
+    const progressStructure = buildProgressInApplication({
```
I think we should also check for responsesLoading and return if it's true, (would need to add it to watched variables array)

andreievg reviewed (changes requested):
Missed a comment for logic, still have to look at ProgressBar UI

andreievg commented:
@nmadruga @CarlosNZ. 

There was a conversation above about clickable pages, not sure the result of it. 

This is something we can ask Debbie about, but I think we should still have sections and pages highlightable as if you can click them, even if canNavigate is false, and show a message when trying to navigate to them ? Can be a generic message for linear "Can only move to this page if requirements of previous page are fullfilled"

CarlosNZ commented:
Could I just make one final UI comment if it's not too late? I find it quite disconcerting not being able to tell (visually) what page I'm on within a Section. Can the current page show slightly different in the Sidebar (maybe bold or indented?) so there's some indication of the current page. With sections, you've made the current section show in a bigger size type, and that works great as a visual indicator.

andreievg commented:
@CarlosNZ 

> Could I just make one final UI comment if it's not too late? I find it quite disconcerting not being able to tell (visually) what page I'm on within a Section. Can the current page show slightly different in the Sidebar (maybe bold or indented?) so there's some indication of the current page. With sections, you've made the current section show in a bigger size type, and that works great as a visual indicator.

https://github.com/openmsupply/application-manager-web-app/pull/82#discussion_r520942397

I think the only remaining thing si to make sure the header is moved to be a pageNumber as text inside list item

nmadruga commented on src/utils/types.ts:
```diff
@@ -94,10 +101,33 @@ interface TemplateSectionPayload {
+
+type ProgressStatus = 'VALID' | 'NOT_VALID' | 'INCOMPLETE'
+
+type ReviewCode = 'PR'
```
Hmmm, that is just to create the structure to be sent to the `Progress bar`, so each section has it's code (to be used in the URL when the user clicks the section/page link) and to create the structure with a `Summary And Review` section I had to add some code to it. Can add a more descriptive name to it.

nmadruga replied:
I can't think of a better solution for now, so have simply changed to use a constant to be used for as the summary section code. Check [commit](https://github.com/openmsupply/application-manager-web-app/pull/127/commits/85cd6e55e5f398d3b9d3f769ac33e2ef9a7cd0ef)

andreievg replied:
I think that works for now, at least no confusion now what it's about

nmadruga commented:
> There was a conversation above about clickable pages, not sure the result of it.

For now the current page is disabled and in linear application all next pages are disable was well.

> This is something we can ask Debbie about, but I think we should still have sections and pages highlightable as if you can click them, even if canNavigate is false, and show a message when trying to navigate to them ?

I have created an issue to change the pages to be all enabled and show the message to users: #142

nmadruga commented:
> I think the only remaining thing si to make sure the header is moved to be a pageNumber as text inside list item

Attempt to address that [here](https://github.com/openmsupply/application-manager-web-app/pull/127/commits/6b4c963c10c0de86f5665f6bb220c0759729a260)

andreievg commented on src/containers/Application/ApplicationPageWrapper.tsx:
```diff
@@ -111,4 +167,84 @@ function processRedirect(appState: any): void {
+      pageName: `Page ${pageNumber}`,
+      canNavigate: isLinear ? isPreviousPage(section.index, pageNumber) : true,
+      isActive: isCurrentPage(pageNumber),
+      status: isCurrentPage(pageNumber)
```
I still don't quite understand the need for condition like this here.

If you forget about `strict` mode setting in validationPage -> it just return valid, invalid or incomplete.

Then the logic seem quite simple to me:
```
getStatue = () => {
draftPageStatus = validatePage
if(isLinear || strict) return draftPageStatus === valid : valid ? not valid

return draftPageStatus
}
```

And for canNavigate
```
canNavigate = true
...
let sectionsStructure: ProgressInApplication = templateSections.map((section) => {
...
const pages = pageNumbers.map((pageNumber) => {
...
status = getPageStatus()
returnObject = {
...
status,
canNavigate
...
}
...
canNavigate = status !== not valid
...
return returnObject
}

```

andreievg commented on src/utils/helpers/validatePage.tsx:
```diff
@@ -0,0 +1,91 @@
+      return [
+        ...responsesStatuses,
+        {
+          status: isValid ? PROGRESS_STATUS.VALID : PROGRESS_STATUS.NOT_VALID,
```
I don't see any returns for incomplete, I still think the logic here can be further simplified with suggested from here https://github.com/openmsupply/application-manager-web-app/pull/127#discussion_r529871184

andreievg reviewed (changes requested):
Further changes please, looking better though

nmadruga commented:
@andreievg I have added the changes as you suggested after our talk. I have updated the `validatePage` describing what is the validation considering, and I think the statuses are looking better now with more `INCOMPLETE` status on pages. Hopefully this is good to go. I think any further changes might be better done in separated PR, just because this one is becoming too big, which isn't helpful.

andreievg commented on src/containers/Application/ApplicationPageWrapper.tsx:
```diff
@@ -111,4 +168,79 @@ function processRedirect(appState: any): void {
+      status:
+        section.index === currentSection
+          ? PROGRESS_STATUS.INCOMPLETE
+          : getCombinedStatus(pages.map(({ status }) => status)),
```
nice

andreievg reviewed (approved):
It's looking good, thanks @nmadruga for all the changes, sorry for all the requests.

I still want to go through with testing and make an issue for extra bits.

nmadruga commented on src/utils/helpers/validatePage.tsx:
```diff
@@ -0,0 +1,78 @@
+      const { text, isValid } = responses[code] as ResponseFull
+      if (isRequired && (!text || isValid === null))
+        return [...responsesStatuses, PROGRESS_STATUS.INCOMPLETE]
+      if (!text || isValid === null) return [...responsesStatuses, PROGRESS_STATUS.VALID]
```
Added [here](https://github.com/openmsupply/application-manager-web-app/pull/127/commits/6870ea4e3c11586120e7796752b1748b507c51f7)

nmadruga commented on src/containers/Application/ApplicationPageWrapper.tsx:
```diff
@@ -111,4 +168,79 @@ function processRedirect(appState: any): void {
+    const progressInSection: ProgressInSection = {
+      code: section.code,
+      title: section.title,
+      canNavigate: isLinear ? section.index <= currentSection : true,
```
Yes, good point

nmadruga replied:
Made a change to try to address this [here](https://github.com/openmsupply/application-manager-web-app/pull/127/commits/7e8a56e9dfa2ed3b1222dcce8c689afec6d28cd7)
